tiny_dnn/core/kernels/fully_connected_op_avx.h

```diff
@@ -14,14 +14,21 @@ namespace kernels {
 
 #ifdef CNN_USE_AVX
 
-template <typename Allocator>
-inline void avx_fully_connected_forward_kernel(
-  const std::vector<std::vector<float, Allocator>> &in_data,
-  const std::vector<float, Allocator> &W,
-  const std::vector<float, Allocator> &bias,
-  std::vector<std::vector<float, Allocator>> &out_data,
-  const fully_params &params,
-  const bool layer_parallelize) {
+template <class E1,
+          class E2,
+          class E3,
+          class E4,
+          value_is_float<E1> * = nullptr,
+          are_all_xexpr<E1, E2, E3, E4> * = nullptr>
+inline void avx_fully_connected_forward_kernel(const E1 &in_data,
+                                               const E2 &W,
+                                               const E3 &bias,
+                                               E4 &out_data,
+                                               const fully_params &params,
+                                               const bool layer_parallelize) {
```

Comment:
I was going to use it for perfect forwarding as advised by @SylvainCorlay . Another option is to use the patter proposed upper in the thread. I chose this one since it's less invasive and shorter (doesn't involve multiple `e1.derived_cast();`)